setup.py

```diff
@@ -29,16 +29,19 @@
 import errno
 import platform
 from traceback import print_exc
+try:
+    import cffi
+except ImportError:
+    cffi = None
 
 # whether any kind of bdist is happening
 # do this before importing anything from distutils
 doing_bdist = any(arg.startswith('bdist') for arg in sys.argv[1:])
 
-if any(bdist in sys.argv for bdist in ['sdist', 'bdist_wheel', 'bdist_egg']):
-    import setuptools
+from setuptools import setup
```

Comment:
any supported python should be able to use setuptools, and I think is to be preferred over distutils where possible

Reply:
If we're switching to requiring setuptools, let's make sure to avoid the main reason we haven't required setuptools: that it breaks `python setup.py install` by implicitly doing egg installs and corrupting sys.path:

```python
from setuptools.command.bdist_egg import bdist_egg

class bdist_egg_disabled(bdist_egg):
    """Disabled version of bdist_egg

    Prevents setup.py install from performing setuptools' default easy_install,
    which it should never ever do.
    """

    def run(self):
        sys.exit(
            "Aborting implicit building of eggs. Use `pip install .` to install from source."
        )

...
setup_args['cmdclass'] = {
    ...
    'bdist_egg': bdist_egg if 'bdist_egg' in sys.argv else bdist_egg_disabled,
}
```